text/0000-event-handler.md

```diff
@@ -0,0 +1,132 @@
+Being able to call to JavaScript from any thread would be useful for binding authors trying to bind native libraries:
+- with custom threading models
+- with progress state
+- which produce events
+- with callbacks
+- ...
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC introduces a new struct `neon::event::EventHandler` that is clone-able and can be send across threads.    
+The user provides a `function` (and an optional `this` context which defaults to the global object) to the struct constructors.
+To actually invoke JavaScript code the `schedule` method is used. This method accepts a closure which will receive a `neon::Context` and the `this`/`function` pair provided during the creation of `neon::event::EventHandler`.
+
+Example for providing the current progress of a background operation:
+
+```rust
+    let mut this = cx.this();
+    let func = cx.argument::<JsFunction>(0)?;
+    let cb = EventHandler::new(func);
+    // or
+    let cb = EventHandler::bind(this, func);
+    thread::spawn(move || {
+        for i in 0..100 {
+            // do some work ....
+            thread::sleep(Duration::from_millis(40));
+            // schedule a call into javascript
+            cb.schedule(move |cx| {
+                // return the arguments of the function call
+                vec![cx.number(i).upcast()]
+            }
+        }
+    });
+```
+
+Here the `EventHandler` "captures" `this` and `func` and calls the closure from the JavaScript thread with the context (`cx`). The values returned by the closure are used as arguments of the function call.
+
+*Note:* The closure is send to the main Javascript thread so every captured value will be moved.   
+
+This approach should be familiar to every Rust programmer as it is the same as `std::thread::spawn` uses.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The struct is implemented as following:
+
+```rust
+struct EventHandlerInner(*mut c_void);
+
+unsafe impl Send for EventHandlerInner {}
+unsafe impl Sync for EventHandlerInner {}
+
+impl Drop for EventHandlerInner {
+    fn drop(&mut self) {
+        // free resources
+    }
+}
+
+#[derive(Clone)]
+pub struct EventHandler(Arc<EventHandlerInner>);
+
+impl EventHandler {
+    // creates a new EventHandler with global as this
+    pub fn new(function: Handle<JsFunction>) -> Self;
+    // creates a new EventHandler with a custom this
+    pub fn bind<T: Value>(this: Handle<T>, function: Handle<JsFunction>) -> Self;
+
+    // schedule a js function call with the arguments returned from the closure
+    pub fn schedule<T, F>(&self, closure: F)
+        where T: Value
+              F: FnOnce(&mut TaskContext, Handle<JsValue>, Handle<JsFunction>) -> Vec<Handle<T>>,
```

Comment:
Probably `IntoIterator` can be used instead of `Vec`, but that can be decided during the implementation

Reply:
It needs to be turned into a `Vec` for other methods. Making it `IntoIterator` would be inefficient for the common case of already having a vector.

I really like this idea, but I think for it to make sense _everything_ that takes a `Vec<Handle<JsValue>>` should be switched to `IntoIterator`.

Reply:
@kjvalencik `Vec` already implements `IntoIterator`, so it can be passed directly.